src/opm/input/eclipse/Schedule/MSW/WellSegments.cpp

```diff
@@ -112,175 +115,249 @@ namespace Opm {
-                                  double volume,
-                                  bool data_ready) {
-
-       Segment segment(segment_number, branch, outlet_segment, length, depth, internal_diameter, roughness, cross_area, volume, data_ready);
-       const int segment_index = this->segmentNumberToIndex(segment_number);
-       if (segment_index < 0) {
-           segment_number_to_index[segment_number] = this->size();
-           this->m_segments.push_back( std::move(segment) );
-       } else
-           this->m_segments[segment_index] = std::move(segment);
-   }
-
-
-    void WellSegments::loadWELSEGS( const DeckKeyword& welsegsKeyword ) {
-
-        // for the first record, which provides the information for the top segment
-        // and information for the whole segment set
+    void WellSegments::addSegment(const Segment& new_segment)
+    {
+        // Decide whether to push_back or insert
+        const auto segment_number = new_segment.segmentNumber();
+        const auto segment_index = segmentNumberToIndex(segment_number);
+        if (segment_index < 0) {
+            // New segment object.
+            const auto new_index = static_cast<int>(this->size());
+            this->segment_number_to_index.insert_or_assign(segment_number, new_index);
+            this->m_segments.push_back(new_segment);
+        }
+        else {
+            // Update to existing segment object.
+            this->m_segments[segment_index] = new_segment;
+        }
+    }
+
+    void WellSegments::addSegment(const int segment_number,
+                                  const int branch,
+                                  const int outlet_segment,
+                                  const double length,
+                                  const double depth,
+                                  const double internal_diameter,
+                                  const double roughness,
+                                  const double cross_area,
+                                  const double volume,
+                                  const bool data_ready,
+                                  const double node_x,
+                                  const double node_y)
+    {
+        auto segment = Segment {
+            segment_number, branch, outlet_segment,
+            length, depth, internal_diameter, roughness,
+            cross_area, volume,
+            data_ready, node_x, node_y
+        };
+        const int segment_index = this->segmentNumberToIndex(segment_number);
+        if (segment_index < 0) {
+            // New segment object.
+            const auto new_index = static_cast<int>(this->size());
+            this->segment_number_to_index.insert_or_assign(segment_number, new_index);
+            this->m_segments.push_back(std::move(segment));
+        }
+        else {
+            // Update to existing segment object.
+            this->m_segments[segment_index] = std::move(segment);
+        }
```

Comment:
Seems like this is the same code as in the other addSegment method (modulo the std::move which in never got). Maybe factor it out into its own method and reuse to prevent further deviations in the future?